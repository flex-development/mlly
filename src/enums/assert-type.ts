/**
 * @file Enums - AssertType
 * @module mlly/enums/AssertType
 */

/**
 * `import` assertion types.
 *
 * ::: info
 * It's unclear whether or not the HTML spec will require assertion types for
 * WebAssembly modules.
 * :::
 *
 * @see https://github.com/WebAssembly/esm-integration/issues/42
 *
 * @enum {Lowercase<string>}
 */
enum AssertType {
  IMPLICIT = 'javascript',
  JSON = 'json'
}

export default AssertType
