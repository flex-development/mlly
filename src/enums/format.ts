/**
 * @file Enums - Format
 * @module mlly/enums/Format
 */

/**
 * Module formats.
 *
 * @enum {Lowercase<string>}
 */
enum Format {
  BUILTIN = 'builtin',
  COMMONJS = 'commonjs',
  JSON = 'json',
  MODULE = 'module',
  WASM = 'wasm'
}

export default Format
