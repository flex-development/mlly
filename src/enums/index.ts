/**
 * @file Entry Point - Enums
 * @module mlly/enums
 */

export { default as Format } from './format'
export { default as SpecifierKind } from './kind-specifier'
export { default as StatementKind } from './kind-statement'
export { default as SyntaxKind } from './kind-syntax'
