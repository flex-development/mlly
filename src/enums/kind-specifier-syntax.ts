/**
 * @file Enums - SpecifierSyntaxKind
 * @module mlly/enums/SpecifierSyntaxKind
 */

/**
 * Module specifier syntax kinds.
 *
 * @enum {Lowercase<string>}
 */
enum SpecifierSyntaxKind {
  DYNAMIC = 'dynamic',
  STATIC = 'static'
}

export default SpecifierSyntaxKind
