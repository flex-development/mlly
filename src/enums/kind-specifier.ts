/**
 * @file Enums - SpecifierKind
 * @module mlly/enums/SpecifierKind
 */

/**
 * Module specifier types.
 *
 * @see https://nodejs.org/api/esm.html#terminology
 *
 * @enum {Lowercase<string>}
 */
enum SpecifierKind {
  ABSOLUTE = 'absolute',
  BARE = 'bare',
  RELATIVE = 'relative'
}

export default SpecifierKind
