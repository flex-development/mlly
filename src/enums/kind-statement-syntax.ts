/**
 * @file Enums - StatementSyntaxKind
 * @module mlly/enums/StatementSyntaxKind
 */

/**
 * CommonJS and ESM statement syntax kinds.
 *
 * @enum {Lowercase<string>}
 */
enum StatementSyntaxKind {
  DECLARATION = 'declaration',
  DEFAULT = 'default',
  DEFAULT_WITH_NAMED = 'default-with-named',
  DEFAULT_WITH_NAMESPACE = 'default-with-namespace',
  DYNAMIC = 'dynamic',
  LIST = 'list',
  NAMED = 'named',
  NAMESPACE = 'namespace',
  REQUIRE = 'require',
  SIDE_EFFECT = 'side-effect'
}

export default StatementSyntaxKind
