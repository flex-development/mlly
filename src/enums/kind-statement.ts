/**
 * @file Enums - StatementKind
 * @module mlly/enums/StatementKind
 */

/**
 * CommonJS and ESM statement kinds.
 *
 * @enum {Lowercase<string>}
 */
enum StatementKind {
  EXPORT = 'export',
  IMPORT = 'import',
  REQUIRE = 'require'
}

export default StatementKind
