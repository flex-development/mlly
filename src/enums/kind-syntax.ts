/**
 * @file Enums - SyntaxKind
 * @module mlly/enums/SyntaxKind
 */

/**
 * CommonJS and ESM statement syntax kinds.
 *
 * @enum {Lowercase<string>}
 */
enum SyntaxKind {
  DECLARATION = 'declaration',
  DEFAULT = 'default',
  DEFAULT_WITH_NAMED = 'default-with-named',
  DEFAULT_WITH_NAMESPACE = 'default-with-namespace',
  DYNAMIC = 'dynamic',
  LIST = 'list',
  NAMED = 'named',
  NAMESPACE = 'namespace',
  REQUIRE = 'require',
  SIDE_EFFECT = 'side-effect'
}

export default SyntaxKind
