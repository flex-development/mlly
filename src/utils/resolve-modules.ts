/**
 * @file resolveModules
 * @module mlly/utils/resolveModules
 */

import { SpecifierSyntaxKind } from '#src/enums'
import type { ResolveModuleOptions } from '#src/interfaces'
import type { URL } from 'node:url'
import extractStatements from './extract-statements'
import resolveModule from './resolve-module'

/**
 * Converts all module specifiers in `code` to absolute specifiers.
 *
 * ::: info
 * Useful for converting code to [`data:` URLs][1].
 * :::
 *
 * [1]: https://nodejs.org/api/esm.html#data-imports
 *
 * @see {@linkcode ResolveModuleOptions}
 * @see https://nodejs.org/api/esm.html#terminology
 *
 * @async
 *
 * @param {string} code - Code to evaluate
 * @param {ResolveModuleOptions} [options={}] - Module resolution options
 * @return {Promise<string>} `code` with module specifiers fully resolved
 */
const resolveModules = async (
  code: string,
  options: ResolveModuleOptions = {}
): Promise<string> => {
  for (const statement of extractStatements(code)) {
    // do nothing if statement does not have specifier
    if (!statement.specifier) continue

    // ignore statements with dynamic specifiers
    if (statement.specifier_syntax === SpecifierSyntaxKind.DYNAMIC) continue

    /**
     * Resolved module URL.
     *
     * @const {URL} url
     */
    const url: URL = await resolveModule(statement.specifier, options)

    // replace original specifier
    code = code.replace(
      statement.code,
      statement.code.replace(statement.specifier, url.href)
    )
  }

  return code
}

export default resolveModules
