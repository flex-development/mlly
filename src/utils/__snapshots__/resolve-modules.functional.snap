// Vitest Snapshot v1

exports[`functional:utils/resolveModules > should resolve modules in code 1`] = `
"import isBuiltin from '@flex-development/is-builtin'
import pathe from '@flex-development/pathe'
import { moduleResolve, type ErrnoException } from 'import-meta-resolve'
import { pathToFileURL, URL } from 'node:url'
import type { ResolveOptions as Options } from '../interfaces/index.mjs'
import type { ModuleSpecifierType } from '../types/index.mjs'
import toBareSpecifier from './to-bare-specifier.mjs'
import toRelativeSpecifier from './to-relative-specifier.mjs'

await import(foo)"
`;
